File: L7_Brackets.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

public class Solution
{
    public static void Main(string[] args)
    {
        String s = "{[()()]}";

        Console.WriteLine(solution(s));
    }

    public static int solution(String S)
    {
        var stack = new Stack<int>();

        foreach (var bracket in S)
        {
            switch (bracket)
            {
                case '(':
                    stack.Push(0);
                    break;
                case '[':
                    stack.Push(1);
                    break;
                case '{':
                    stack.Push(2);
                    break;

                case ')':
                    if (stack.Count == 0 || stack.Pop() != 0)
                    {
                        return 0;
                    }
                    break;
                case ']':
                    if (stack.Count == 0 || stack.Pop() != 1)
                    {
                        return 0;
                    }
                    break;
                case '}':
                    if (stack.Count == 0 || stack.Pop() != 2)
                    {
                        return 0;
                    }
                    break;
            }
        }

        if (stack.Count == 0)
        {
            return 1;
        }

        return 0;
    }
}

File: L7_Brackets.py
def solution(S):
    if len(S) % 2 == 1:   return 0

    matched = {"]":"[", "}":"{", ")": "("}
    to_push = ["[", "{", "("]
    stack = []

    for element in S:
        if element in to_push:
            stack.append(element)
        else:
            if len(stack) == 0:
                return 0
            elif matched[element] != stack.pop():
                return 0

    if len(stack) == 0:
        return 1
    else:
        return 0

S = "{[()()]}"
print(solution(S))
